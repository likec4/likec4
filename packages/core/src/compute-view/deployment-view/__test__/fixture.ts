import { map, prop } from 'remeda'
import type { DeploymentRulesBuilderOp } from '../../../builder'
import { Builder } from '../../../builder'
import { LikeC4Model } from '../../../model'
import type { DeploymentView } from '../../../types'
import { withReadableEdges } from '../../utils/with-readable-edges'
import { computeDeploymentView } from '../compute'

const {
  builder: b,
  model: {
    model,
    person,
    system,
    component,
    webapp,
    mobile,
    ...$m
  },
  deployment: {
    deployment,
    env,
    node,
    zone,
    instanceOf,
    ...$d
  },
  views: {
    views,
    deploymentView,
    $rules,
    $include,
    $exclude,
  },
} = Builder.forSpecification({
  elements: {
    person: {
      style: {
        shape: 'person',
      },
    },
    system: {},
    component: {},
    webapp: {
      style: {
        shape: 'browser',
      },
    },
    mobile: {
      style: {
        shape: 'mobile',
      },
    },
  },
  deployments: {
    env: {},
    zone: {},
    node: {},
  },
  tags: ['old', 'next', 'temp'],
  relationships: { 'https': {} },
})

export const builder = b
  .with(
    model(
      person('customer', {
        title: 'Happy Customer',
      }),
      system('cloud').with(
        component('frontend').with(
          webapp('dashboard'),
          mobile('mobile'),
        ),
        component('auth'),
        component('backend').with(
          component('api'),
        ),
        component('media', {
          shape: 'storage',
        }),
      ),
      system('aws').with(
        component('rds', {
          shape: 'storage',
        }),
        component('s3', {
          shape: 'storage',
        }),
      ),
      system('email'),
    ),
    model(
      $m.rel('customer', 'cloud', 'uses services'),
      $m.rel('customer', 'cloud.frontend.mobile', 'opens mobile app'),
      $m.rel('customer', 'cloud.frontend.dashboard', {
        title: 'opens in browser',
        tags: ['old'],
      }),
      $m.rel('cloud.frontend.dashboard', 'cloud.auth', {
        title: 'authenticates',
        color: 'green',
      }),
      $m.rel('cloud.frontend.dashboard', 'cloud.backend.api', {
        title: 'fetches data',
        tags: ['old'],
      }),
      $m.rel('cloud.frontend.dashboard', 'cloud.media', {
        title: 'fetches media',
        tags: ['temp'],
      }),
      $m.rel('cloud.frontend.mobile', 'cloud.auth', {
        title: 'authenticates',
        color: 'amber',
      }),
      $m.rel('cloud.frontend.mobile', 'cloud.backend.api', 'fetches data'),
      $m.rel('cloud.frontend.mobile', 'cloud.media', 'fetches media'),
      $m.rel('cloud.backend.api', 'cloud.auth', 'authorizes'),
      $m.rel('cloud.backend.api', 'cloud.media', 'uploads media'),
      $m.rel('cloud.backend.api', 'aws.rds', 'reads/writes'),
      $m.rel('cloud.backend.api', 'email', 'sends emails'),
      $m.rel('cloud.media', 'aws.s3', 'uploads'),
      $m.rel('email', 'customer', 'sends emails'),
    ),
    deployment(
      node('customer').with(
        instanceOf('instance', 'customer'),
      ),
      env('prod').with(
        zone('eu').with(
          zone('zone1').with(
            instanceOf('api', 'cloud.backend.api'),
            instanceOf('ui', 'cloud.frontend.dashboard'),
          ),
          zone('zone2').with(
            instanceOf('api', 'cloud.backend.api'),
            instanceOf('ui', 'cloud.frontend.dashboard'),
          ),
          instanceOf('auth', 'cloud.auth'),
          instanceOf('media', 'cloud.media'),
          instanceOf('db', 'aws.rds'),
        ),
        zone('us').with(
          zone('zone1').with(
            instanceOf('api', 'cloud.backend.api'),
            instanceOf('ui', 'cloud.frontend.dashboard'),
          ),
          instanceOf('db', 'aws.rds'),
        ),
      ),
      env('dev').with(
        node('devCustomer').with(
          instanceOf('instance', 'customer'),
        ),
        node('devCloud').with(
          instanceOf('instance', 'cloud'),
        ),
      ),
      env('acc').with(
        node('testCustomer').with(
          instanceOf('instance', 'customer'),
        ),
        zone('eu').with(
          instanceOf('api', 'cloud.backend.api'),
          instanceOf('ui', 'cloud.frontend.dashboard'),
          instanceOf('auth', 'cloud.auth'),
          instanceOf('db', 'aws.rds'),
        ),
      ),
      env('global').with(
        instanceOf('email', 'email'),
      ),
      $d.rel('prod.eu.db', 'prod.us.db', {
        title: 'replicates',
        color: 'green',
      }),
      $d.rel('prod.us.db', 'prod.eu.db', {
        title: 'replicates',
        color: 'amber',
      }),
    ),
  )
  .clone()

export type Types = typeof builder['Types']

export { $exclude, $include }

export function computeView(...rules: DeploymentRulesBuilderOp<Types>[]) {
  const modelsource = builder.clone()
    .with(
      views(
        deploymentView('index', $rules(...rules)),
      ),
    ).build()

  const model = LikeC4Model.create({
    ...modelsource,
    views: {},
  })

  const view = withReadableEdges(computeDeploymentView(model, modelsource.views.index as DeploymentView))

  return Object.assign(view, {
    nodeIds: map(view.nodes, prop('id')) as string[],
    edgeIds: map(view.edges, prop('id')) as string[],
  })
}

export function computeNodesAndEdges(...rules: DeploymentRulesBuilderOp<Types>[]) {
  const { nodeIds, edgeIds } = computeView(...rules)
  return {
    // Starts with capital letter to be first in snapshot
    Nodes: nodeIds,
    edges: edgeIds,
  }
}

export function createModel() {
  const modelsource = builder.build()

  return LikeC4Model.create({
    ...modelsource,
    views: {},
  })
}
