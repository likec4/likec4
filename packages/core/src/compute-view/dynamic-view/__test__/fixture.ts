import { partition } from 'remeda'
import type {
  DynamicViewIncludeRule,
  DynamicViewRule,
  DynamicViewStep,
  ElementViewPredicate,
  Fqn,
  ParsedDynamicView as DynamicView,
  ViewId,
} from '../../../types'
import { type $Aux, type FakeElementIds, fakeModel } from '../../element-view/__test__/fixture'
import { computeDynamicView } from '../compute'

const emptyView = {
  _type: 'dynamic' as const,
  id: 'index' as ViewId,
  title: null,
  description: null,
  tags: null,
  links: null,
  rules: [],
}

type StepExpr = `${FakeElementIds} ${'->' | '<-'} ${FakeElementIds}`
type StepProps = Omit<DynamicViewStep, 'source' | 'target' | 'isBackward'>

export function $step(expr: StepExpr, props?: string | Partial<StepProps>): DynamicViewStep {
  const title = typeof props === 'string' ? props : props?.title
  if (expr.includes(' -> ')) {
    const [source, target] = expr.split(' -> ')
    return {
      source: source as Fqn,
      target: target as Fqn,
      ...(typeof props === 'object' ? props : {}),
      title: title ?? null,
    }
  }
  if (expr.includes(' <- ')) {
    const [target, source] = expr.split(' <- ')
    return {
      source: source as Fqn,
      target: target as Fqn,
      ...(typeof props === 'object' ? props : {}),
      title: title ?? null,
      isBackward: true,
    }
  }
  throw new Error(`Invalid step expression: ${expr}`)
}

export function compute(
  stepsAndRules: (DynamicViewStep<$Aux> | ElementViewPredicate<$Aux> | DynamicViewIncludeRule<$Aux>)[],
) {
  const [steps, rules] = partition(stepsAndRules, (s): s is DynamicViewStep => 'source' in s)
  let view = computeDynamicView(
    fakeModel,
    {
      ...emptyView,
      steps,
      rules: rules as DynamicViewRule[],
    } as DynamicView,
  )
  return Object.assign(view, {
    nodeIds: view.nodes.map((node) => node.id) as string[],
    edgeIds: view.edges.map((edge) => edge.id) as string[],
  })
}
