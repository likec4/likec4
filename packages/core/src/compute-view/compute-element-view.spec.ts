import { describe, expect, it } from 'vitest'
import { fakeModel, type FakeElementIds } from '../__test__'
import type { ComputedView, ElementKind, Fqn, Tag, ViewID, ViewRule } from '../types'
import { computeElementView } from './compute-element-view'
import { pluck } from 'rambdax'

const ids = pluck('id')

const emptyView = {
  title: null,
  description: null,
  tags: null,
  links: null,
  rules: []
}

describe('compute-element-view', () => {
  function computeView(...args: [FakeElementIds, ViewRule[]] | [ViewRule[]]) {
    let result: ComputedView
    if (args.length === 1) {
      result = computeElementView(
        {
          ...emptyView,
          id: 'index' as ViewID,
          title: '',
          rules: args[0]
        },
        fakeModel()
      )
    } else {
      result = computeElementView(
        {
          ...emptyView,
          id: 'index' as ViewID,
          title: '',
          viewOf: args[0] as Fqn,
          rules: args[1]
        },
        fakeModel()
      )
    }
    return Object.assign(result, {
      nodeIds: ids(result.nodes),
      edgeIds: ids(result.edges)
    })
  }

  it('should be empty if no root and no rules', () => {
    const { nodes, edges } = computeElementView(
      {
        ...emptyView,
        id: 'index' as ViewID
      },
      fakeModel()
    )
    expect(nodes).toEqual([])
    expect(edges).toEqual([])
  })

  it('should show only root if no rules', () => {
    const { nodes, edges } = computeElementView(
      {
        ...emptyView,
        id: 'index' as ViewID,
        title: '',
        viewOf: 'cloud' as Fqn,
        rules: []
      },
      fakeModel()
    )

    expect(ids(nodes)).toEqual(['cloud'])

    expect(edges).toEqual([])
  })

  it('should return landscape view on top `include *`', () => {
    const { nodes, edges } = computeElementView(
      {
        ...emptyView,
        id: 'index' as ViewID,
        title: '',
        rules: [
          {
            isInclude: true,
            exprs: [
              {
                wildcard: true
              }
            ]
          }
        ]
      },
      fakeModel()
    )
    expect(ids(nodes)).toEqual(['support', 'customer', 'cloud', 'amazon'])
    expect(ids(edges)).to.have.members(['cloud:amazon', 'customer:cloud', 'support:cloud'])
  })

  it('should return landscape view on top `include *, -> cloud.*`', () => {
    const { nodes, edges } = computeElementView(
      {
        ...emptyView,
        id: 'index' as ViewID,
        title: '',
        rules: [
          {
            isInclude: true,
            exprs: [
              { wildcard: true },
              {
                incoming: {
                  element: 'cloud' as Fqn,
                  isDescedants: true
                }
              }
            ]
          }
        ]
      },
      fakeModel()
    )
    expect(ids(nodes)).toEqual(['support', 'customer', 'cloud', 'amazon', 'cloud.frontend'])
    expect(ids(edges)).to.have.same.members([
      'cloud:amazon',
      'customer:cloud.frontend',
      'support:cloud.frontend'
    ])
  })

  it('view of cloud', () => {
    const view = computeElementView(
      {
        ...emptyView,
        id: 'cloud' as ViewID,
        title: '',
        viewOf: 'cloud' as Fqn,
        rules: [
          {
            isInclude: true,
            exprs: [
              {
                wildcard: true
              }
            ]
          }
        ]
      },
      fakeModel()
    )
    const { nodes, edges } = view

    expect(ids(nodes)).toEqual([
      'support',
      'customer',
      'cloud',
      'cloud.frontend',
      'cloud.backend',
      'amazon'
    ])

    expect(ids(edges)).to.have.same.members([
      'cloud.frontend:cloud.backend',
      'cloud.backend:amazon',
      'customer:cloud.frontend',
      'support:cloud.frontend'
    ])

    expect(view).toMatchSnapshot()
  })

  it('view of cloud.backend', () => {
    const view = computeElementView(
      {
        ...emptyView,
        id: 'cloudbackend' as ViewID,
        title: '',
        viewOf: 'cloud.backend' as Fqn,
        rules: [
          {
            isInclude: true,
            exprs: [
              {
                wildcard: true
              },
              {
                element: 'customer' as Fqn,
                isDescedants: false
              }
            ]
          }
        ]
      },
      fakeModel()
    )
    const { nodes, edges } = view

    expect(ids(nodes)).toEqual([
      'customer',
      'cloud.frontend',
      'cloud.backend',
      'cloud.backend.graphql',
      'cloud.backend.storage',
      'amazon'
    ])

    expect(ids(edges)).to.have.same.members([
      'cloud.backend.graphql:cloud.backend.storage',
      'cloud.backend.storage:amazon',
      'cloud.frontend:cloud.backend.graphql',
      'customer:cloud.frontend'
    ])
  })

  it('view of cloud.frontend', () => {
    const view = computeElementView(
      {
        ...emptyView,
        id: 'cloudfrontend' as ViewID,
        title: '',
        viewOf: 'cloud.frontend' as Fqn,
        rules: [
          {
            isInclude: true,
            exprs: [
              {
                wildcard: true
              }
            ]
          }
        ]
      },
      fakeModel()
    )
    const { nodes, edges } = view

    expect(ids(nodes)).toEqual([
      'support',
      'customer',
      'cloud.frontend',
      'cloud.frontend.adminPanel',
      'cloud.frontend.dashboard',
      'cloud.backend'
    ])

    expect(ids(edges)).to.have.same.members([
      'cloud.frontend.adminPanel:cloud.backend',
      'cloud.frontend.dashboard:cloud.backend',
      'customer:cloud.frontend.dashboard',
      'support:cloud.frontend.adminPanel'
    ])

    expect(view).toMatchSnapshot()
  })

  it('view of cloud.frontend (and include parent cloud)', () => {
    const view = computeElementView(
      {
        ...emptyView,
        id: 'cloudfrontend2' as ViewID,
        title: '',
        viewOf: 'cloud.frontend' as Fqn,
        rules: [
          {
            isInclude: true,
            exprs: [
              {
                wildcard: true
              },
              // including parent cloud should not remove implicit cloud.backend
              {
                element: 'cloud' as Fqn,
                isDescedants: false
              }
            ]
          }
        ]
      },
      fakeModel()
    )
    const { nodes, edges } = view

    expect(ids(nodes)).toEqual([
      'support',
      'customer',
      'cloud',
      'cloud.frontend',
      'cloud.frontend.adminPanel',
      'cloud.frontend.dashboard',
      'cloud.backend'
    ])

    expect(ids(edges)).to.have.same.members([
      'cloud.frontend.adminPanel:cloud.backend',
      'cloud.frontend.dashboard:cloud.backend',
      'customer:cloud.frontend.dashboard',
      'support:cloud.frontend.adminPanel'
    ])
  })

  it('view of cloud (exclude cloud, amazon.*)', () => {
    const view = computeElementView(
      {
        ...emptyView,
        id: 'cloud' as ViewID,
        title: '',
        viewOf: 'cloud' as Fqn,
        rules: [
          {
            isInclude: true,
            exprs: [
              {
                wildcard: true
              }
            ]
          },
          {
            isInclude: false,
            exprs: [
              { element: 'cloud' as Fqn, isDescedants: false },
              { element: 'amazon' as Fqn, isDescedants: true }
            ]
          }
        ]
      },
      fakeModel()
    )
    const { nodes, edges } = view

    expect(ids(nodes)).toEqual(['support', 'customer', 'cloud.frontend', 'cloud.backend'])

    expect(ids(edges)).to.have.same.members([
      'cloud.frontend:cloud.backend',
      'customer:cloud.frontend',
      'support:cloud.frontend'
    ])

    expect(view).toMatchSnapshot()
  })

  it('view with 3 levels', () => {
    const view = computeElementView(
      {
        ...emptyView,
        id: 'cloud3levels' as ViewID,
        title: '',
        viewOf: 'cloud' as Fqn,
        rules: [
          {
            isInclude: true,
            exprs: [
              // include *
              { wildcard: true },
              // include cloud.frontend.*
              { element: 'cloud.frontend' as Fqn, isDescedants: true },
              // include cloud.backend.*
              { element: 'cloud.backend' as Fqn, isDescedants: true }
            ]
          },
          {
            isInclude: false,
            exprs: [
              // exclude cloud.frontend
              { element: 'cloud.frontend' as Fqn, isDescedants: false }
            ]
          }
        ]
      },
      fakeModel()
    )

    expect(view.nodes.map(n => n.id)).toEqual([
      'support',
      'customer',
      'cloud',
      'cloud.frontend.adminPanel',
      'cloud.frontend.dashboard',
      'cloud.backend',
      'cloud.backend.graphql',
      'cloud.backend.storage',
      'amazon'
    ])

    expect(view).toMatchSnapshot()
  })

  it('view of amazon', () => {
    const { nodes, edges } = computeElementView(
      {
        ...emptyView,
        id: 'amazon' as ViewID,
        title: '',
        viewOf: 'amazon' as Fqn,
        rules: [
          {
            isInclude: true,
            exprs: [
              // include *
              { wildcard: true },
              // include cloud
              { element: 'cloud' as Fqn, isDescedants: false },
              // include cloud.* -> amazon
              {
                source: { element: 'cloud' as Fqn, isDescedants: true },
                target: { element: 'amazon' as Fqn, isDescedants: false }
              }
            ]
          }
        ]
      },
      fakeModel()
    )

    expect(ids(nodes)).toEqual(['cloud', 'cloud.backend', 'amazon', 'amazon.s3'])

    expect(ids(edges)).to.have.same.members(['cloud.backend:amazon.s3'])
  })

  it('index view with applied styles', () => {
    const { nodes } = computeView([
      {
        isInclude: true,
        exprs: [{ wildcard: true }, { element: 'cloud.frontend' as Fqn, isDescedants: false }]
      },
      // all elements
      // color: secondary
      {
        targets: [{ wildcard: true }],
        style: {
          color: 'secondary',
          shape: 'storage'
        }
      },
      // cloud
      // color: muted
      {
        targets: [{ element: 'cloud' as Fqn, isDescedants: false }],
        style: {
          color: 'muted'
        }
      },
      // cloud.*
      // shape: browser
      {
        targets: [{ element: 'cloud' as Fqn, isDescedants: true }],
        style: {
          shape: 'browser'
        }
      }
    ])

    const amazon = nodes.find(n => n.id === 'amazon')!
    const customer = nodes.find(n => n.id === 'customer')!
    const cloud = nodes.find(n => n.id === 'cloud')!
    const frontend = nodes.find(n => n.id === 'cloud.frontend')!

    expect(amazon).toMatchObject({
      color: 'secondary',
      shape: 'storage'
    })
    expect(customer).toMatchObject({
      color: 'secondary',
      shape: 'storage'
    })

    expect(cloud).toMatchObject({
      color: 'muted',
      shape: 'storage'
    })
    expect(frontend).toMatchObject({
      parent: 'cloud',
      color: 'secondary',
      shape: 'browser'
    })
  })

  it('should include by element kind', () => {
    const { nodeIds, edgeIds } = computeView([
      {
        isInclude: true,
        exprs: [
          {
            elementKind: 'system' as ElementKind,
            isEqual: true
          }
        ]
      }
    ])

    expect(nodeIds).toEqual(['cloud', 'amazon'])
    expect(edgeIds).toEqual(['cloud:amazon'])
  })

  it('should include by element tag', () => {
    const { nodeIds, edgeIds } = computeView([
      {
        isInclude: true,
        exprs: [
          {
            elementTag: 'old' as Tag,
            isEqual: true
          }
        ]
      }
    ])

    expect(nodeIds).toEqual(['cloud.backend.storage', 'cloud.frontend.adminPanel'])

    expect(edgeIds).toEqual([])
  })

  it('should exclude by element tag and kind', () => {
    const { nodes, edges } = computeView('cloud', [
      {
        isInclude: true,
        exprs: [
          // include *
          { wildcard: true },
          // include cloud.backend.*
          { element: 'cloud.backend' as Fqn, isDescedants: true },
          // include cloud.frontend.*
          { element: 'cloud.frontend' as Fqn, isDescedants: true }
        ]
      },
      {
        isInclude: false,
        exprs: [
          {
            elementKind: 'actor' as ElementKind,
            isEqual: true
          },
          {
            elementTag: 'old' as Tag,
            isEqual: true
          }
        ]
      }
    ])

    expect(ids(nodes)).toEqual([
      'cloud',
      'cloud.frontend',
      'cloud.frontend.dashboard',
      'cloud.backend',
      'amazon',
      'cloud.backend.graphql'
    ])

    expect(ids(edges)).to.have.same.members([
      'cloud.frontend.dashboard:cloud.backend.graphql',
      'cloud.frontend:cloud.backend.graphql',
      'cloud.backend:amazon'
    ])
  })
})
