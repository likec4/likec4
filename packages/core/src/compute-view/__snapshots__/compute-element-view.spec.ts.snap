// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compute-element-view > view of cloud (exclude cloud, amazon.*) 1`] = `
{
  "autoLayout": "TB",
  "description": null,
  "edges": [
    {
      "id": "cloud.frontend:cloud.backend",
      "label": null,
      "parent": null,
      "relations": [
        "cloud.frontend.dashboard:cloud.backend.graphql",
        "cloud.frontend.adminPanel:cloud.backend.graphql",
      ],
      "source": "cloud.frontend",
      "target": "cloud.backend",
    },
    {
      "id": "support:cloud.frontend",
      "label": null,
      "parent": null,
      "relations": [
        "support:cloud.frontend.adminPanel",
      ],
      "source": "support",
      "target": "cloud.frontend",
    },
    {
      "id": "customer:cloud.frontend",
      "label": null,
      "parent": null,
      "relations": [
        "customer:cloud.frontend.dashboard",
      ],
      "source": "customer",
      "target": "cloud.frontend",
    },
  ],
  "id": "cloud",
  "links": null,
  "nodes": [
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "support",
      "kind": "actor",
      "links": null,
      "parent": null,
      "shape": "person",
      "tags": null,
      "technology": null,
      "title": "support",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "customer",
      "kind": "actor",
      "links": null,
      "parent": null,
      "shape": "person",
      "tags": null,
      "technology": null,
      "title": "customer",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.frontend",
      "kind": "container",
      "links": null,
      "parent": null,
      "shape": "browser",
      "tags": null,
      "technology": null,
      "title": "frontend",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.backend",
      "kind": "container",
      "links": null,
      "parent": null,
      "shape": "rectangle",
      "tags": null,
      "technology": null,
      "title": "backend",
    },
  ],
  "rules": [
    {
      "exprs": [
        {
          "wildcard": true,
        },
      ],
      "isInclude": true,
    },
    {
      "exprs": [
        {
          "element": "cloud",
          "isDescedants": false,
        },
        {
          "element": "amazon",
          "isDescedants": true,
        },
      ],
      "isInclude": false,
    },
  ],
  "tags": null,
  "title": "",
  "viewOf": "cloud",
}
`;

exports[`compute-element-view > view of cloud 1`] = `
{
  "autoLayout": "TB",
  "description": null,
  "edges": [
    {
      "id": "cloud.frontend:cloud.backend",
      "label": null,
      "parent": "cloud",
      "relations": [
        "cloud.frontend.dashboard:cloud.backend.graphql",
        "cloud.frontend.adminPanel:cloud.backend.graphql",
      ],
      "source": "cloud.frontend",
      "target": "cloud.backend",
    },
    {
      "id": "cloud.backend:amazon",
      "label": null,
      "parent": null,
      "relations": [
        "cloud.backend.storage:amazon.s3",
      ],
      "source": "cloud.backend",
      "target": "amazon",
    },
    {
      "id": "support:cloud.frontend",
      "label": null,
      "parent": null,
      "relations": [
        "support:cloud.frontend.adminPanel",
      ],
      "source": "support",
      "target": "cloud.frontend",
    },
    {
      "id": "customer:cloud.frontend",
      "label": null,
      "parent": null,
      "relations": [
        "customer:cloud.frontend.dashboard",
      ],
      "source": "customer",
      "target": "cloud.frontend",
    },
  ],
  "id": "cloud",
  "links": null,
  "nodes": [
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "support",
      "kind": "actor",
      "links": null,
      "parent": null,
      "shape": "person",
      "tags": null,
      "technology": null,
      "title": "support",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "customer",
      "kind": "actor",
      "links": null,
      "parent": null,
      "shape": "person",
      "tags": null,
      "technology": null,
      "title": "customer",
    },
    {
      "children": [
        "cloud.frontend",
        "cloud.backend",
      ],
      "color": "primary",
      "description": null,
      "id": "cloud",
      "kind": "system",
      "links": null,
      "parent": null,
      "shape": "rectangle",
      "tags": null,
      "technology": null,
      "title": "cloud",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.frontend",
      "kind": "container",
      "links": null,
      "parent": "cloud",
      "shape": "browser",
      "tags": null,
      "technology": null,
      "title": "frontend",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.backend",
      "kind": "container",
      "links": null,
      "parent": "cloud",
      "shape": "rectangle",
      "tags": null,
      "technology": null,
      "title": "backend",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "amazon",
      "kind": "system",
      "links": null,
      "parent": null,
      "shape": "rectangle",
      "tags": null,
      "technology": null,
      "title": "amazon",
    },
  ],
  "rules": [
    {
      "exprs": [
        {
          "wildcard": true,
        },
      ],
      "isInclude": true,
    },
  ],
  "tags": null,
  "title": "",
  "viewOf": "cloud",
}
`;

exports[`compute-element-view > view of cloud.frontend 1`] = `
{
  "autoLayout": "TB",
  "description": null,
  "edges": [
    {
      "id": "cloud.frontend.adminPanel:cloud.backend",
      "label": null,
      "parent": null,
      "relations": [
        "cloud.frontend.adminPanel:cloud.backend.graphql",
      ],
      "source": "cloud.frontend.adminPanel",
      "target": "cloud.backend",
    },
    {
      "id": "cloud.frontend.dashboard:cloud.backend",
      "label": null,
      "parent": null,
      "relations": [
        "cloud.frontend.dashboard:cloud.backend.graphql",
      ],
      "source": "cloud.frontend.dashboard",
      "target": "cloud.backend",
    },
    {
      "id": "support:cloud.frontend.adminPanel",
      "label": null,
      "parent": null,
      "relations": [
        "support:cloud.frontend.adminPanel",
      ],
      "source": "support",
      "target": "cloud.frontend.adminPanel",
    },
    {
      "id": "customer:cloud.frontend.dashboard",
      "label": null,
      "parent": null,
      "relations": [
        "customer:cloud.frontend.dashboard",
      ],
      "source": "customer",
      "target": "cloud.frontend.dashboard",
    },
  ],
  "id": "cloudfrontend",
  "links": null,
  "nodes": [
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "support",
      "kind": "actor",
      "links": null,
      "parent": null,
      "shape": "person",
      "tags": null,
      "technology": null,
      "title": "support",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "customer",
      "kind": "actor",
      "links": null,
      "parent": null,
      "shape": "person",
      "tags": null,
      "technology": null,
      "title": "customer",
    },
    {
      "children": [
        "cloud.frontend.adminPanel",
        "cloud.frontend.dashboard",
      ],
      "color": "primary",
      "description": null,
      "id": "cloud.frontend",
      "kind": "container",
      "links": null,
      "parent": null,
      "shape": "browser",
      "tags": null,
      "technology": null,
      "title": "frontend",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.frontend.adminPanel",
      "kind": "component",
      "links": null,
      "parent": "cloud.frontend",
      "shape": "rectangle",
      "tags": [
        "old",
      ],
      "technology": null,
      "title": "adminPanel",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.frontend.dashboard",
      "kind": "component",
      "links": null,
      "parent": "cloud.frontend",
      "shape": "rectangle",
      "tags": null,
      "technology": null,
      "title": "dashboard",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.backend",
      "kind": "container",
      "links": null,
      "parent": null,
      "shape": "rectangle",
      "tags": null,
      "technology": null,
      "title": "backend",
    },
  ],
  "rules": [
    {
      "exprs": [
        {
          "wildcard": true,
        },
      ],
      "isInclude": true,
    },
  ],
  "tags": null,
  "title": "",
  "viewOf": "cloud.frontend",
}
`;

exports[`compute-element-view > view with 3 levels 1`] = `
{
  "autoLayout": "TB",
  "description": null,
  "edges": [
    {
      "id": "cloud.frontend.adminPanel:cloud.backend.graphql",
      "label": null,
      "parent": "cloud",
      "relations": [
        "cloud.frontend.adminPanel:cloud.backend.graphql",
      ],
      "source": "cloud.frontend.adminPanel",
      "target": "cloud.backend.graphql",
    },
    {
      "id": "cloud.frontend.dashboard:cloud.backend.graphql",
      "label": null,
      "parent": "cloud",
      "relations": [
        "cloud.frontend.dashboard:cloud.backend.graphql",
      ],
      "source": "cloud.frontend.dashboard",
      "target": "cloud.backend.graphql",
    },
    {
      "id": "cloud.backend.graphql:cloud.backend.storage",
      "label": null,
      "parent": "cloud.backend",
      "relations": [
        "cloud.backend.graphql:cloud.backend.storage",
      ],
      "source": "cloud.backend.graphql",
      "target": "cloud.backend.storage",
    },
    {
      "id": "cloud.backend.storage:amazon",
      "label": null,
      "parent": null,
      "relations": [
        "cloud.backend.storage:amazon.s3",
      ],
      "source": "cloud.backend.storage",
      "target": "amazon",
    },
    {
      "id": "support:cloud.frontend.adminPanel",
      "label": null,
      "parent": null,
      "relations": [
        "support:cloud.frontend.adminPanel",
      ],
      "source": "support",
      "target": "cloud.frontend.adminPanel",
    },
    {
      "id": "customer:cloud.frontend.dashboard",
      "label": null,
      "parent": null,
      "relations": [
        "customer:cloud.frontend.dashboard",
      ],
      "source": "customer",
      "target": "cloud.frontend.dashboard",
    },
  ],
  "id": "cloud3levels",
  "links": null,
  "nodes": [
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "support",
      "kind": "actor",
      "links": null,
      "parent": null,
      "shape": "person",
      "tags": null,
      "technology": null,
      "title": "support",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "customer",
      "kind": "actor",
      "links": null,
      "parent": null,
      "shape": "person",
      "tags": null,
      "technology": null,
      "title": "customer",
    },
    {
      "children": [
        "cloud.frontend.adminPanel",
        "cloud.frontend.dashboard",
        "cloud.backend",
      ],
      "color": "primary",
      "description": null,
      "id": "cloud",
      "kind": "system",
      "links": null,
      "parent": null,
      "shape": "rectangle",
      "tags": null,
      "technology": null,
      "title": "cloud",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.frontend.adminPanel",
      "kind": "component",
      "links": null,
      "parent": "cloud",
      "shape": "rectangle",
      "tags": [
        "old",
      ],
      "technology": null,
      "title": "adminPanel",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.frontend.dashboard",
      "kind": "component",
      "links": null,
      "parent": "cloud",
      "shape": "rectangle",
      "tags": null,
      "technology": null,
      "title": "dashboard",
    },
    {
      "children": [
        "cloud.backend.graphql",
        "cloud.backend.storage",
      ],
      "color": "primary",
      "description": null,
      "id": "cloud.backend",
      "kind": "container",
      "links": null,
      "parent": "cloud",
      "shape": "rectangle",
      "tags": null,
      "technology": null,
      "title": "backend",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.backend.graphql",
      "kind": "component",
      "links": null,
      "parent": "cloud.backend",
      "shape": "rectangle",
      "tags": null,
      "technology": null,
      "title": "graphql",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.backend.storage",
      "kind": "component",
      "links": null,
      "parent": "cloud.backend",
      "shape": "rectangle",
      "tags": [
        "old",
      ],
      "technology": null,
      "title": "storage",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "amazon",
      "kind": "system",
      "links": null,
      "parent": null,
      "shape": "rectangle",
      "tags": null,
      "technology": null,
      "title": "amazon",
    },
  ],
  "rules": [
    {
      "exprs": [
        {
          "wildcard": true,
        },
        {
          "element": "cloud.frontend",
          "isDescedants": true,
        },
        {
          "element": "cloud.backend",
          "isDescedants": true,
        },
      ],
      "isInclude": true,
    },
    {
      "exprs": [
        {
          "element": "cloud.frontend",
          "isDescedants": false,
        },
      ],
      "isInclude": false,
    },
  ],
  "tags": null,
  "title": "",
  "viewOf": "cloud",
}
`;
